src/compiler/checker.ts

```diff
@@ -31515,6 +31515,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 }
             }
 
+            if (findAncestor(right.parent, n => isDecorator(n) && isClassLike(n.parent))) {
+                grammarErrorOnNode(right, Diagnostics.Private_identifiers_are_not_allowed_outside_class_bodies);
+                return anyType;
+            }

```

Comment:
I think this incorrectly issues an error on the following code:

```ts
class C {
    #myProp = "hello!";
    method() {
        @decorator(() => this.#myProp)
        class C {}

        return C;
    }
}

new C().method();


function decorator(f: () => string): <T>(x: T, ctxt: any) => T {
    console.log(f());
    return (x, _ctxt) => x;
}
```

@rbuckton can confirm.

Reply:
Yes, a class decorator should only have access to private identifiers defined in the same lexical scope but *outside* of the class it decorats.

Reply:
Classes offer tremendous flexibility 😄. What are your thoughts on using this approach to detect the scope of private fields? Perhaps there is a more efficient alternative/existing helper...

Reply:
Hm - I kind of wish we could just do this check when a private field fails to resolve - but then it wouldn't be a grammar check.

Reply:
Then again, we provide no error on the following JavaScript

```js
class C {
    method() { this.#wat }
}
```

Whereas most engines do error on this. So in a sense it's weird that we don't issue a *grammar* error on an invalid private name access.

That said, I'm not sure if we want to go there.